packages/insomnia/src/ui/components/modals/login-modal.tsx

```diff
@@ -1,132 +1,236 @@
+    }
   }
 
   hide() {
-    this.modal?.hide();
+    this.modalRef.current?.hide();
   }
+}
 
-  render() {
-    const { title, message, loading, error } = this.state;
-    return (
-      <form onSubmit={this._handleLogin} data-testid="LoginModal__form">
-        <Modal ref={this._setModalRef} {...this.props}>
-          <ModalHeader>{title || 'Log Into Your Account'}</ModalHeader>
-          <ModalBody className="pad">
-            {message ? <p className="notice info">{message}</p> : null}
-            <div className="form-control form-control--outlined no-pad-top">
-              <label>
-                Email
-                <input
-                  type="email"
-                  // @ts-expect-error -- TSCONVERSION appears to be genuine
-                  required="required"
-                  placeholder="me@mydomain.com"
-                  ref={this._setEmailInputRef}
-                />
-              </label>
-            </div>
-            <div className="form-control form-control--outlined">
-              <label>
-                Password
-                <input
-                  type="password"
-                  // @ts-expect-error -- TSCONVERSION appears to be genuine
-                  required="required"
-                  placeholder="•••••••••••••••••"
-                  ref={this._setPasswordInputRef}
-                />
-              </label>
-            </div>
-            {error ? <div className="danger pad-top">** {error}</div> : null}
-          </ModalBody>
-          <ModalFooter>
-            <div className="margin-left">
-              Don't have an account yet? <Link href="https://app.insomnia.rest">Sign Up</Link>
-            </div>
-            <button type="submit" className="btn">
-              {loading ? <i className="fa fa-spin fa-refresh margin-right-sm" /> : null}
-              Log In
+export const LoginModal = memo(forwardRef<LoginModalHandle, {}>(function LoginModal({ ...props }, ref) {
+  const modalRef = useRef<Modal>(null);
+  const tokenInputRef = useRef<HTMLInputElement>(null);
+  const urlInputRef = useRef<HTMLInputElement>(null);
+
+  const [state, setState] = useState<State & Options>({
+    state: 'ready',
+    error: '',
+    title: '',
+    message: '',
+    reauth: false,
+    url: '',
+  });
+
+  useImperativeHandle(ref, () => {
+    currentLoginModalHandle = new LoginModalHandle(modalRef, setState);
+    return currentLoginModalHandle;
+  }, []);
+
+  const reset = useCallback(() => {
+    setState(state => ({ ...state, state: 'ready', error: '' }));
+  }, []);
+
+  const goToTokenEntry = useCallback(() => {
+    setState(state => ({ ...state, state: 'token-entry' }));
+  }, []);
+
+  const submitToken = useCallback((event: FormEvent) => {
+    event.preventDefault();
+    currentLoginModalHandle?.submitAuthCode(tokenInputRef.current?.value ?? '');
+  }, []);
+
+  const copyUrl = useCallback(() => {
+    urlInputRef.current?.select();
+    document.execCommand('copy');
+  }, []);
```

Comment:
issue: this implementation of copying broke for me when I tried it, causing me to not be able to log in.

he below fixes it (including the `import { clipboard } from 'electron'`):


```suggestion
  const copyUrl = useCallback(() => {
    clipboard.writeText(state.url);
  }, [state.url]);
```


My reproduction steps were:
- clicked `Log In` in the Account dropdown
- it redirects to a new open tab in a Chrome workspace that I didn't want it to (i.e. one that I don't have access to my insomnia password anyway)
- so I went back and clicked "Copy URL" and it did not copy anything

Also, document.execCommand is [deprecated](https://developer.mozilla.org/en-US/docs/Web/API/Document/execCommand) anyway.  Check out `copy-button.tsx` to see us using the clipboard export from electron in the same way.

Also, this implementation is dependent on the selection, whereas the suggested fix is not.  When first trying to figure out what was broken I tried selecting from the input, but it kept just copying `loginKey` to my clipboard.  Later I realized when I looked at the implementation that it was because of the selection being used, rather than the url.

Reply:
Thanks, but: you didn't need to write all of this; It was only a mistake. I copied the same code from the browser portion, which has no choice but to do this, because browsers have intentionally limited clipboard APIs.

Reply:
(also fixed)